orca-clouddriver/src/main/groovy/com/netflix/spinnaker/orca/clouddriver/pipeline/providers/gce/GceDeployStagePreProcessor.java

```diff
@@ -107,26 +114,25 @@ private static boolean shouldPinSourceServerGroup(String strategy) {
     return Strategy.fromStrategyKey(strategy) == Strategy.RED_BLACK;
   }
 
-  private Map<String, Object> getResizeContext(StageData stageData) {
-    AbstractDeployStrategyStage.CleanupConfig cleanupConfig =
-        AbstractDeployStrategyStage.CleanupConfig.fromStage(stageData);
-    Map<String, Object> resizeContext = new HashMap<>();
-
-    resizeContext.put(
-        cleanupConfig.getLocation().singularType(), cleanupConfig.getLocation().getValue());
-    resizeContext.put("cluster", cleanupConfig.getCluster());
-    resizeContext.put("moniker", cleanupConfig.getMoniker());
-    resizeContext.put("credentials", cleanupConfig.getAccount());
-    resizeContext.put("cloudProvider", cleanupConfig.getCloudProvider());
-
-    ResizeStrategy.Source source = getSource(targetServerGroupResolver, stageData, resizeContext);
-    resizeContext.put("serverGroupName", source.getServerGroupName());
-    resizeContext.put("action", ResizeStrategy.ResizeAction.scale_to_server_group);
-    resizeContext.put("source", source);
-    resizeContext.put(
-        "useNameAsLabel", true); // hint to deck that it should _not_ override the name
-
-    return resizeContext;
+  private Optional<Map<String, Object>> getResizeContext(StageData stageData) {
```

Comment:
FYI: I want to (and am planning on to do this this week) to remove this logic form the preprocessor. because it's a) confusing and not explicit and b) can cause issues...
the specific issue is, imagine you have:

```
v1 server group - disabled
v0 server group - enabled
```

now you run the deploy and expect this to pin `v0`, but in (at least AWS case today) this will pin `v1` because the source determination here runs before we run the determinesourceservergroup task.


my plan is to have a "PinSourceSupport" class that the strategy can call explicitly to add the pin/unpin stages - doesn't seem like this should belong to the preprocessor - and also no point for each cloud provider to duplicate this code

Reply:
I so appreciate you reworking this logic and making it easier to understand!

Since for some reason a GCE red/black has a pin stage, but for AWS only a _rolling_ red/black has a pin stage, I think the GCE-specific pre-processor will continue need to add its own pin stage :cry: I have no context on the origin of that discrepancy but I'm hesitant to change it in case GCE users are relying on the current functionality.

Reply:
I am impressed you pin for RB - it's a good idea, but pretty dangerous. We enabled it briefly on AWS and @dreynaud can tell you about the awesome incident it caused... anyway. most of the issues are around "guaranteeing" that the ASG gets unpinned which we don't do a very good job at today

Reply:
Haha can't wait to hear about the incident... I only started looking into this because it recently came to our attention that in GCE, we are basically guaranteeing that the ASG will _never_ be unpinned correctly because we are snapshotting its capacity _after_ pinning it :face_with_head_bandage: (stay tuned for a fix for that in an upcoming PR)

Reply:
I did a twitter webposting on it! It _will_ sound familiar to you 😅  https://twitter.com/FakeRyanGosling/status/1106714429247221761

Reply:
Amazing write-up! Love that it took y'all 24 hours to fix it and I think it will have taken us on the order of years to fix for GCE :see_no_evil: :see_no_evil: :see_no_evil: 

Excuse me while I LOL at your other tweets for the rest of the evening...